pip/pip-418.md

```diff
@@ -0,0 +1,174 @@
+
+When BookKeeper is not the default storage, the behaviors are not clear for these components unless looking into the code. This proposal aims at a clear definition of the behavior.
+
+# Goals
+
+## In Scope
+
+Define the behavior for components that rely on BookKeeper when BookKeeper is not the default storage.
+
+# High Level Design
+
+When BookKeeper is not the default storage, 
+
+| Component | Behavior |
+| :- | :- |
+| Schema Registry | Fail fast |
+| Delayed Delivery Tracker | Fail fast if it's `BucketDelayedDeliveryTrackerFactory` |
+| Compaction Service | Fallback to a dummy implementation |
+| `/bookies/all` REST API | Return an empty bookie list |
+| `PersistentTopic#getInternalStats` | Return an empty ledger list in stats |
+
+Here we don't fail fast for the compaction service because whether to perform compaction does not affect the correctness. It only affects the storage size.
+
+# Detailed Design
+
+## Design & Implementation Details
+
+Simplify the `ManagedLedgerStorage#initialize` method by removing the `BookKeeperClientFactory` and `EventLoopGroup` parameters:
+
+```java
+public interface ManagedLedgerStorage extends AutoCloseable {
+
+    void initialize(ServiceConfiguration conf,
+                    MetadataStoreExtended metadataStore,
+                    OpenTelemetry openTelemetry) throws Exception;
+```
+
+The reason to have these parameters is that `BookKeeperClientFactory` is maintained in `PulsarService` and creating a `BookKeeper` instance requires a `EventLoopGroup` object.
+
+However, these BookKeeper related parameters are not useful for a `ManagedLedgerStorage` implementation. Even if it's another BookKeeper based implementation, it should create its own BookKeeper client objects according to the configuration (`conf`). In addition, the `EventLoopGroup` instance passed to the method represents the `pulsar-io` thread pool, which is coupled with Pulsar's internal details. The custom implementation should create is own thread pool.
+
+Since the built-in implementation still requires an `EventLoopGroup` instance, we can skip calling `initialize` and pass the arguments directly to the constructor: 
+
+```java
+if (config.getManagedLedgerStorageClassName().equals(ManagedLedgerClientFactory.class.getName())) {
+    return new ManagedLedgerClientFactory(config, localMetadataStore, ioEventLoopGroup, openTelemetry);
+} else {
+    // create() will pass these arguments to initialize()
+    return ManagedLedgerStorage.create(config, localMetadataStore, openTelemetry);
+}
+```
+
+Add a new method to `BookkeeperManagedLedgerStorageClass`:
+
+```java
+public interface BookkeeperManagedLedgerStorageClass extends ManagedLedgerStorageClass {
+    /* ... */
+    BookKeeperClientFactory getBookKeeperClientFactory();
```

Comment:
Is it designed to be used by `BookkeeperSchemaStorageFactory` and `BookkeeperBucketSnapshotStorage`?

I mean, BookkeeperManagedLedgerStorageClass already has `getBookKeeperClient`, and should all Pulsar's internal implementations use the same bookkeeper client? Maybe I missed something.